protocol/x/clob/types/untriggered_conditional_orders.go

```diff
@@ -0,0 +1,83 @@
+package types
+
+import "bytes"
+
+// MinConditionalOrderHeap is type alias for `[]Order` which implements "container/heap"
+// interface.
+//
+// This is a _MIN_ heap. Orders are compared by their `ConditionalOrderTriggerSubticks` field, and then by
+// their `OrderHash` if the trigger subticks are equal.
+type MinConditionalOrderHeap []Order
+
+func (h MinConditionalOrderHeap) Len() int {
+	return len(h)
+}
+
+func (h MinConditionalOrderHeap) Less(i, j int) bool {
+	x, y := h[i], h[j]
+
+	// If the trigger subticks are the same, sort by order hash.
+	// This is required for determinism in the case of multiple orders with the same trigger subticks.
+	if x.ConditionalOrderTriggerSubticks == y.ConditionalOrderTriggerSubticks {
+		xHash := x.GetOrderHash()
+		yHash := y.GetOrderHash()
+		return bytes.Compare(xHash[:], yHash[:]) == -1
+	}
+	return x.ConditionalOrderTriggerSubticks < y.ConditionalOrderTriggerSubticks
+}
+
+func (h MinConditionalOrderHeap) Swap(i, j int) {
+	h[i], h[j] = h[j], h[i]
+}
+
+func (h *MinConditionalOrderHeap) Push(x interface{}) {
+	*h = append(*h, x.(Order))
+}
+
+func (h *MinConditionalOrderHeap) Pop() interface{} {
+	old := *h
+	n := len(old)
+	x := old[n-1]
+	*h = old[0 : n-1]
+	return x
+}
+
+// MaxConditionalOrderHeap is type alias for `[]Order` which implements "container/heap"
+// interface.
+//
+// This is a _MAX_ heap. Orders are compared by their `ConditionalOrderTriggerSubticks` field, and then by
+// their `OrderHash` if the trigger subticks are equal.
+type MaxConditionalOrderHeap []Order
+
+func (h MaxConditionalOrderHeap) Len() int {
+	return len(h)
+}
+
+func (h MaxConditionalOrderHeap) Less(i, j int) bool {
+	x, y := h[i], h[j]
+
+	// If the trigger subticks are the same, sort by order hash.
+	// This is required for determinism in the case of multiple orders with the same trigger subticks.
+	if x.ConditionalOrderTriggerSubticks == y.ConditionalOrderTriggerSubticks {
+		xHash := x.GetOrderHash()
+		yHash := y.GetOrderHash()
+		return bytes.Compare(xHash[:], yHash[:]) == -1
+	}
+	return x.ConditionalOrderTriggerSubticks > y.ConditionalOrderTriggerSubticks
+}
+
+func (h MaxConditionalOrderHeap) Swap(i, j int) {
+	h[i], h[j] = h[j], h[i]
+}
+
+func (h *MaxConditionalOrderHeap) Push(x interface{}) {
+	*h = append(*h, x.(Order))
+}
+
+func (h *MaxConditionalOrderHeap) Pop() interface{} {
+	old := *h
+	n := len(old)
+	x := old[n-1]
+	*h = old[0 : n-1]
```

Comment:
the official documentation for `container/heap` has some safety checks here for memory:
```
func (pq *PriorityQueue) Pop() any {
	old := *pq
	n := len(old)
	item := old[n-1]
	old[n-1] = nil  // avoid memory leak
	item.index = -1 // for safety
	*pq = old[0 : n-1]
	return item
}
```

Should we do this?

Reply:
Also, when we poll the order we look at the 0th element in the queue, as seen [here](https://github.com/dydxprotocol/v4-chain/pull/996/files#diff-fc56c9715292486eb043611b0cfee5372bc9ca67921ffac388d175e504b4ac01R254). However, for both the min and max heap we are returning the last element in the array (old[n-1]). Is that correct? Don't we want to poll/pop the same element?

Reply:
For your second comment, I was similarly confused.

The confusing part is that `heap.Pop(h)` does not simply call `h.Pop()` under the hood. Whereas `h.Pop()` returns the lowest priority element, `heap.Pop(h)` returns the highest priority element. This is what is extremely confusing. The logic of `heap.Pop(h)` is inessence:
1. Call `h.Pop()` to remove the lowest priority element
1. Replace the highest priority element with the popped element
1. Call `heap.Fix(h, 0)` to re-sort the replaced item
1. Return the original highest-priority element that you removed

[Here's](https://cs.opensource.google/go/go/+/master:src/container/heap/heap.go) the source code in case you want to read line-by-line and understand.

Reply:
You could consider adding an `h.Peek()` method which simply returns `h[0]`